eth-types/src/sign_types.rs

```diff
@@ -0,0 +1,149 @@
+    /// Secp256k1 public key
+    pub pk: Secp256k1Affine,
+    /// Hash of the message that is being signed
+    pub msg_hash: secp256k1::Fq,
+}
+
+lazy_static! {
+    static ref SIGN_DATA_DEFAULT: SignData = {
+        let generator = Secp256k1Affine::generator();
+        let sk = secp256k1::Fq::one();
+        let pk = generator * sk;
+        let pk = pk.to_affine();
+        let msg_hash = secp256k1::Fq::one();
+        let randomness = secp256k1::Fq::one();
+        let (sig_r, sig_s) = sign(randomness, sk, msg_hash);
+
+        SignData {
+            signature: (sig_r, sig_s),
+            pk,
+            msg_hash,
+        }
+    };
+}
+
+impl Default for SignData {
+    fn default() -> Self {
+        // Hardcoded valid signature corresponding to a hardcoded private key and
+        // message hash generated from "nothing up my sleeve" values to make the
+        // ECDSA chip pass the constraints, to be use for padding signature
+        // verifications (where the constraints pass, but we don't care about the
+        // message hash and public key).
+        SIGN_DATA_DEFAULT.clone()
+    }
+}
+
+/// Convert a `BigUint` into 32 bytes in little endian.
+pub fn biguint_to_32bytes_le(v: BigUint) -> [u8; 32] {
+    let mut res = [0u8; 32];
+    let v_le = v.to_bytes_le();
+    res[..v_le.len()].copy_from_slice(&v_le);
+    res
+}
+
+/// Recover the public key from a secp256k1 signature and the message hash.
+pub fn recover_pk(
+    v: u8,
+    r: &Word,
+    s: &Word,
+    msg_hash: &[u8; 32],
+) -> Result<Secp256k1Affine, libsecp256k1::Error> {
+    let mut sig_bytes = [0u8; 64];
+    sig_bytes[..32].copy_from_slice(&r.to_be_bytes());
+    sig_bytes[32..].copy_from_slice(&s.to_be_bytes());
+    let signature = libsecp256k1::Signature::parse_standard(&sig_bytes)?;
+    let msg_hash = libsecp256k1::Message::parse_slice(msg_hash.as_slice())?;
+    let recovery_id = libsecp256k1::RecoveryId::parse(v)?;
+    let pk = libsecp256k1::recover(&msg_hash, &signature, &recovery_id)?;
+    let pk_be = pk.serialize();
+    let pk_le = pk_bytes_swap_endianness(&pk_be[1..]);
+    let mut pk_bytes = secp256k1::Serialized::default();
+    pk_bytes.as_mut().copy_from_slice(&pk_le[..]);
+    let pk = Secp256k1Affine::from_bytes(&pk_bytes);
+    ct_option_ok_or(pk, libsecp256k1::Error::InvalidPublicKey)
+}
+
+lazy_static! {
+    /// Secp256k1 Curve Scalar.  Referece: Section 2.4.1 (parameter `n`) in "SEC 2: Recommended
+    /// Elliptic Curve Domain Parameters" document at http://www.secg.org/sec2-v2.pdf
+    pub static ref SECP256K1_Q: BigUint = BigUint::from_slice(&[
+        0xd0364141, 0xbfd25e8c,
+        0xaf48a03b, 0xbaaedce6,
+        0xfffffffe, 0xffffffff,
+        0xffffffff, 0xffffffff,
+    ]);
+}
+
+/// Helper function to convert a `CtOption` into an `Result`.  Similar to
+/// `Option::ok_or`.
+pub fn ct_option_ok_or<T, E>(v: CtOption<T>, err: E) -> Result<T, E> {
+    Option::<T>::from(v).ok_or(err)
+}
```

Comment:
I believe I'm already using the conversion you implemented in this function.  So what I would like to do here is convert a `CtOption<T>` to a `Result<T, E>` with the minimum code, because this is done many times.
How would you do it without using this auxiliary function but keeping it short?

Reply:
My bad here..